torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,195 @@
+import itertools
+from typing_extensions import Protocol
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
```

Comment:
I'd like to motivate this initial forward pass better. Effectively, the module's parameters don't exist before the first forward pass is run, and some PyTorch functionality assumes those parameters already exist (and why wouldn't they?). To initialize the lazy modules, then, a "dry run" with inputs of a "representative size", dtype and device should be performed. 

Detailing the dry run is a little tricky. In PyTorch networks can be very dynamic and have multiple branches. A single initializing dry run may not call foward() on each module in a network. If lazy modules appear in two of a network's branches, for example, then two initializing dry runs would be necessary, and the inputs for each dry run must be constructed to trigger the different paths through the network.

You can try to address this complexity or suggest, for now at least, that users not try to use lazy modules with networks with multiple branches.

Reply:
I tried to add some comments from your reply to this docstring.